### PR TITLE
14115: further convert cleanup

### PR DESCRIPTION
**Pull request by mkalish**
This PR adds some additional logging to the convert step and adds the logs from the convert step into the submission endpoint

Test Steps:
1. Send this message
```sh
curl --location 'http://localhost:7071/api/reports' \
--header 'client: ignore.ignore-mars-otc' \
--header 'Content-Type: application/hl7-v2' \
--data 'MSH|^~\&|MMTC.PROD^2.16.840.1.113883.3.8589.4.2.106.1^ISO|CAREEVOLUTION^00Z0000024^CLIA|AIMS.INTEGRATION.PRD^2.16.840.1.114222.4.3.15.1^ISO|AIMS.PLATFORM^2.16.840.1.114222.4.1.217446^ISO|20240403205305+0000||ORU^R01^ORU_R01|20240403205305_dba7572cc6334f1ea0744c5f235c823e|P|2.5.1|||NE|NE|||||PHLabReport-NoAck^ELR251R1_Rcvr_Prof^2.16.840.1.113883.9.11^ISO
SFT|CAREEVOLUTION|2022|MMTC.PROD|16498||20240402
PID|1||8be6fa3710374dcebe0174e0fd5a1a7c^^^MMTC.PROD&2.16.840.1.113883.3.8589.4.2.106.1&ISO^PI||^^^^^^~^^^^^^||||||^^^^02139^USA||^^^^^111^1111111
ORC|RE||^MMTC.PROD^2.16.840.1.113883.3.8589.4.2.106.1^ISO|||||||||^^||^^^^^^|||||||SA.OTCSelfReport|^^^^02139^^^^|^^^^^^
OBR|1||^MMTC.PROD^2.16.840.1.113883.3.8589.4.2.106.1^ISO|94558-4^SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay^LN^^^^2.71|||20240403120000|||||||||^^|^^^^^^|||||20240403120000|||F
OBX|1|CWE|94558-4^SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay^LN^^^^2.71||260373001^Detected^SCT^^^^20200901||||||F||||00Z0000042||BinaxNOW COVID-19 Antigen Self Test_Abbott Diagnostics Scarborough, Inc._EUA^^99ELR^^^^Vunknown||20240403120000||||SA.OTCSelfReport^^^^^&2.16.840.1.113883.3.8589.4.1.152&ISO^XX^^^00Z0000042|
NTE|1|L|Note
OBX|2|NM|35659-2^Age at specimen collection^LN^^^^2.71||24|a^year^UCUM^^^^2.1|||||F||||00Z0000042||||||||SA.OTCSelfReport^^^^^&2.16.840.1.113883.3.8589.4.1.152&ISO^XX^^^00Z0000042||||||QST
SPM|1|^dba7572cc6334f1ea0744c5f235c823e&MMTC.PROD&2.16.840.1.113883.3.8589.4.2.106.1&ISO||697989009^Anterior nares swab^SCT^^^^20200901|||||||||||||20240403120000-0400|20240403120000-0400'
```
2. Watch the terminal and verify that the logs from the convert step have the report id and topic (look for `"TOPIC":"\"mars-otc-elr\""`)
3. Grab the outputted idea and hit the submission endpoint
```sh
curl --location 'http://localhost:7071/api/waters/report/<REPORTID>/history' \
--header 'authentication-type: okta' \
--header 'Authorization: Bearer dummy'
```
4. Verify that the validation error is in the response `"Item 1 in the report was not valid. Reason: HL7 was not valid at OBX[1]-19[1].1"`

## Changes
- Updates the logging in the convert step
- Properly generates a report if no bundles will be routed
- Updates the submission endpoint to return logs from the convert step

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [x] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #14115 

## To Be Done
https://github.com/CDCgov/prime-reportstream/issues/14350
https://github.com/CDCgov/prime-reportstream/issues/14349

### CODE REVIEW

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt [hunk 1]
```diff
@@ -98,124 +100,147 @@ class FHIRConverter(
...
+                }
+            } else {
+                val nextEvent = ProcessEvent(
+                    Event.EventAction.NONE,
+                    queueMessage.reportId,
+                    Options.None,
+                    emptyMap(),
+                    emptyList()
+                )
+                val report = Report(
```
Comment (mkalish) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt:216]: So this really is an empty report...none of the bundles make it through so it should have zero items.  However, due to some checks in the lineage generation, that code will actually throw an exception if you create a report without any items/and item lineage.

There is tech there to clean for sure, but I'm going to add a ticket and defer that.

File: prime-router/src/main/kotlin/history/SubmissionHistory.kt [hunk 2]
```diff
@@ -410,6 +414,23 @@ class DetailedSubmissionHistory(
...
     }
 
+    /**
+     * Enrich a parent detailed history with details from the convert action.
+     * Add errors, and warnings, to the history details.
+     *
+     * @param descendant route action that will be used to enrich
+     */
+    private fun enrichWithConvertAction(descendant: DetailedSubmissionHistory) {
+        require(
```
Comment (JFisk42) [inline prime-router/src/main/kotlin/history/SubmissionHistory.kt:424]: I was going to ask if we needed to unit test the negative case here, but I think based on my understanding this is technically not possible for those conditions to be falsified in the current setup, right?

**Review by JFisk42 — APPROVED** [review]
Ran the test steps and everything looks good

**Review by arnejduranovic — COMMENTED** [review]
The suggested manual test is great! However, validation error is just one type of error that can occur. Another being "parse" or "FHIR Translation", for example. Do we have tests that verify that all the different types of errors that can occur in the convert step show up correctly in the history endpoint? I'm thinking:

- We can't test these cases manually because the receive step catches them
- These tests can serve as good documentation on what the various error cases are and how they are communicated to a sender

File: prime-router/src/main/kotlin/Report.kt [hunk 3]
```diff
@@ -415,6 +415,53 @@ class Report : Logging {
...
+     * [metadata] is the metadata to use, mocked meta is passed in for testing
+     * [itemLineage] itemlineages for this report to track parent/child reports
+     */
+    constructor(
+        bodyFormat: Format,
+        sources: List<Source>,
+        metadata: Metadata? = null,
+        parentItemLineageData: List<ParentItemLineageData>,
+        destination: Receiver? = null,
+        nextAction: TaskAction = TaskAction.process,
```
[the remark this answers is not in the document]
Comment (mkalish) [reply prime-router/src/main/kotlin/Report.kt:434]: Copy pasta, I'll delete

File: prime-router/src/main/kotlin/Report.kt [hunk 4]
```diff
@@ -415,6 +415,53 @@ class Report : Logging {
...
+        topic: Topic,
+    ) {
+        this.id = UUID.randomUUID()
+        // UP submissions do not need a schema, but it is required by the database to maintain legacy functionality
```
[the remark this answers is not in the document]
Comment (mkalish) [reply prime-router/src/main/kotlin/Report.kt:438]: Super legacy, and just copied over from the other constructor

File: prime-router/src/main/kotlin/Report.kt [hunk 5]
```diff
@@ -415,6 +415,53 @@ class Report : Logging {
         this.nextAction = nextAction
     }
 
+    data class ParentItemLineageData(val parentReportId: UUID, val parentReportIndex: Int)
+
+    /**
+     * Full ELR Report constructor for ingest
+     * [bodyFormat] is the format for this report. Should be HL7
+     * [sources] is the ClientSource or TestSource, where this data came from
+     * [numberOfMessages] how many incoming messages does this Report represent
```
[the remark this answers is not in the document]
Comment (mkalish) [reply prime-router/src/main/kotlin/Report.kt:424]: Yep will update

[the remark this answers is not in the document]
Comment (mkalish) [reply prime-router/src/main/kotlin/Report.kt:424]: Yep I'll update

**Comment by mkalish** [comment]
This needs to get rebased against master before approving.

**Review by arnejduranovic — APPROVED** [review]

**Pull request merged**